### PR TITLE
enable dictionary for custom ds

### PR DESCRIPTION
**Pull request by ianhe8x**
Add `dictionaryQuery: (filter: F) => DictionaryQueryEntry;` to ds's SecondLayerHandlerProcessor interface.

Exclude baseFilter if dictionaryQuery is provided.

### CODE REVIEW

**Comment by ianhe8x** [comment]
@stwiname I tested with tutorials-moonriver-evm-starter, changed startBlock to 1. and it works.
But it still would be good to have you review the changes.

**Review by stwiname — COMMENTED** [review]
I've tested this locally on the tutorial project and didn't see any difference between this an main branch. All i changed was the start block to 1, are any other changes needed?

**Comment by ianhe8x** [comment]
> I've tested this locally on the tutorial project and didn't see any difference between this an main branch. All i changed was the start block to 1, are any other changes needed?

Provide dictionary in the network section `https://api.subquery.network/sq/subquery/moonriver-dictionary`, 
and also use the new built contract-processor

File: packages/contract-processors/src/moonbeam.ts [hunk 1]
```diff
@@ -271,6 +273,33 @@ const EventProcessor: SecondLayerHandlerProcessor<
...
+      return;
+    }
+
+    // Follows bloom filters https://docs.ethers.io/v5/concepts/events/#events--filters
+    if (filter?.topics) {
+      for (let i = 0; i < Math.min(filter.topics.length, 4); i++) {
+        const topic = filter.topics[i];
+        if (!topic) {
+          continue;
+        }
```
[the remark this answers is not in the document]
Comment (ianhe8x) [reply packages/contract-processors/src/moonbeam.ts:293]: it is `continue` here, so the next topic will still be handled and result in a query condition.

[the remark this answers is not in the document]
Comment (ianhe8x) [reply packages/contract-processors/src/moonbeam.ts:293]: The place where i'm not certain is in which case a topic can be represented as `string[]`

[the remark this answers is not in the document]
Comment (ianhe8x) [reply packages/contract-processors/src/moonbeam.ts:292]: @stwiname do you think we can drop the support of bloom filters, it is quite complex and challenging to be supported by dictionary query, which means change TopicFilter to `type TopicFilter = string | null | undefined;`

[the remark this answers is not in the document]
Comment (stwiname) [reply packages/contract-processors/src/moonbeam.ts:284]: I had to understand the dictionary functionality better. I think it is fine to drop support for `or` functionality on the filters, it won't be a common used feature and a PR could always be made to bring it back.

**Review by stwiname — APPROVED** [review]

File: packages/contract-processors/src/moonbeam.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ianhe8x) [reply packages/contract-processors/src/moonbeam.ts:284]: It is `OR` between `queryEntry`, and `AND` between `queryEntry.conditions`

**Pull request merged**